selfdrive/ui/replay/filereader.h

```diff
@@ -1,70 +1,87 @@
 #pragma once
 
-#include <thread>
+#include <unordered_map>
+#include <vector>
 
 #include <QElapsedTimer>
 #include <QMultiMap>
 #include <QNetworkAccessManager>
-#include <QReadWriteLock>
 #include <QString>
-#include <QVector>
-#include <QWidget>
+#include <QThread>
 
-#include <bzlib.h>
 #include <capnp/serialize.h>
-#include <kj/io.h>
-#include "cereal/gen/cpp/log.capnp.h"
 
-#include "tools/clib/channel.h"
+#include "cereal/gen/cpp/log.capnp.h"
 
 class FileReader : public QObject {
   Q_OBJECT
 
 public:
-  FileReader(const QString& file_);
-  void startRequest(const QUrl &url);
-  ~FileReader();
-  virtual void readyRead();
-  void httpFinished();
-  virtual void done() {};
+  FileReader(const QString &fn, QObject *parent = nullptr);
+  void read();
+  void abort();
 
-public slots:
-  void process();
-
-protected:
-  QNetworkReply *reply;
+signals:
+  void finished(const QByteArray &dat);
+  void failed(const QString &err);
 
 private:
-  QNetworkAccessManager *qnam;
-  QElapsedTimer timer;
-  QString file;
+  void startHttpRequest();
+  QNetworkReply *reply_ = nullptr;
+  QUrl url_;
 };
 
-typedef QMultiMap<uint64_t, cereal::Event::Reader> Events;
+enum CameraType {
+  RoadCam = 0,
+  DriverCam,
+  WideRoadCam
+};
+const CameraType ALL_CAMERAS[] = {RoadCam, DriverCam, WideRoadCam};
+const int MAX_CAMERAS = std::size(ALL_CAMERAS);
 
-class LogReader : public FileReader {
-Q_OBJECT
+struct EncodeIdx {
+  int segmentNum;
+  uint32_t frameEncodeId;
+};
+
+class Event {
```

Comment:
think it would be cleaner without this class. we should just use the event reader directly

Reply:
In addition, `which` and `mono_time` will be used  for sorting in  #21004, to avoid the problem that event may be lost or resent if sorted only by mono_time